eth/tracers/api_blockResult.go

```diff
@@ -0,0 +1,347 @@
+		env.state.Prepare(tx.Hash(), i)
+		vmenv := vm.NewEVM(env.blockCtx, core.NewEVMTxContext(msg), env.state, api.backend.ChainConfig(), vm.Config{})
+		if _, err := core.ApplyMessage(vmenv, msg, new(core.GasPool).AddGas(msg.Gas())); err != nil {
+			failed = err
+			break
+		}
+		// Finalize the state so any modifications are written to the trie
+		// Only delete empty objects if EIP158/161 (a.k.a Spurious Dragon) is in effect
+		env.state.Finalise(vmenv.ChainConfig().IsEIP158(block.Number()))
+	}
+	close(jobs)
+	pend.Wait()
+
+	// If execution failed in between, abort
+	if failed != nil {
+		return nil, failed
+	}
+
+	return api.fillBlockResult(env, block)
+}
+
+func (api *API) getTxResult(env *traceEnv, state *state.StateDB, index int, block *types.Block) error {
+	tx := block.Transactions()[index]
+	msg, _ := tx.AsMessage(env.signer, block.BaseFee())
+	from, _ := types.Sender(env.signer, tx)
+	to := tx.To()
+
+	txctx := &Context{
+		BlockHash: block.TxHash(),
+		TxIndex:   index,
+		TxHash:    tx.Hash(),
+	}
+	// Call Prepare to clear out the statedb access list
+	state.Prepare(txctx.TxHash, txctx.TxIndex)
+
+	sender := &types.AccountWrapper{
+		Address:  from,
+		Nonce:    state.GetNonce(from),
+		Balance:  (*hexutil.Big)(state.GetBalance(from)),
+		CodeHash: state.GetCodeHash(from),
+	}
+	var receiver *types.AccountWrapper
+	if to != nil {
+		receiver = &types.AccountWrapper{
+			Address:  *to,
+			Nonce:    state.GetNonce(*to),
+			Balance:  (*hexutil.Big)(state.GetBalance(*to)),
+			CodeHash: state.GetCodeHash(*to),
+		}
+	}
+	// Run the transaction with tracing enabled.
+	vmenv := vm.NewEVM(env.blockCtx, core.NewEVMTxContext(msg), state, api.backend.ChainConfig(), vm.Config{Debug: true, Tracer: env.tracer, NoBaseFee: true})
+	// Computes the new state by applying the given message.
+	env.rMu.Lock()
+	result, err := core.ApplyMessage(vmenv, msg, new(core.GasPool).AddGas(msg.Gas()))
+	if err != nil {
+		return fmt.Errorf("tracing failed: %w", err)
+	}
+	env.rMu.Unlock()
+	// If the result contains a revert reason, return it.
+	returnVal := result.Return()
+	if len(result.Revert()) > 0 {
+		returnVal = result.Revert()
+	}
+
+	createdAcc := env.tracer.CreatedAccount()
+	var after []*types.AccountWrapper
+	if to == nil {
+		if createdAcc == nil {
+			return errors.New("unexpected tx: address for created contract unavailable")
+		}
+		to = &createdAcc.Address
+	}
+	// collect affected account after tx being applied
+	for _, acc := range []common.Address{from, *to, api.backend.Coinbase()} {
+		after = append(after, &types.AccountWrapper{
+			Address:  acc,
+			Nonce:    state.GetNonce(acc),
+			Balance:  (*hexutil.Big)(state.GetBalance(acc)),
+			CodeHash: state.GetCodeHash(acc),
+		})
+	}
+
+	// merge required proof data
+	proofAccounts := env.tracer.UpdatedAccounts()
+	for addr := range proofAccounts {
+		addrStr := addr.String()
+
+		env.pMu.Lock()
+		_, existed := env.Proofs[addrStr]
+		env.pMu.Unlock()
+		if existed {
+			continue
+		}
```

Comment:
What's the rationale here? It seems that we keep the first account proof produced by any of the jobs. But each job will operate on different intermediate states, so this proof might also change.

Reply:
This gets proof after each transaction is executed, so there should be no loss of intermediate state.

Reply:
Shouldn't we call `state.GetProof(addr)` on the base state (the one before the block's execution) and not the intermediate state?